README.md
# CIDR to IPv4 info

### description

Input one CIDR string, and it will help you convert it to more detailed info.

### Example

input: `10.0.0.0/24`

output:

|header|content|
|------|-------|
|`range`|`10.0.0.0/24`|
|`netmask`|`255.255.255.0`|
|`wildcardBits`|`0.0.0.255`|
|`firstIp`|`10.0.0.0`|
|`firstIpInt`|`167772160`|
|`lastIp`|`10.0.0.255`|
|`lastIpInt`|`167772415`|
|`totalHost`|`256`|


CHANGELOG.md
# CIDR Changelog

## [Initial Version] - 2023-05-15

## [First Version] - 2023-06-01

- Finish the full features about converting CIDR to detailed IPv4 information.
- Add icon.
- Add screenshot.
- Add README.

tests/ip.test.ts
import { validateIPv4 } from '../src/utils/common-utils';

describe('ip validation', function () {
    it('10.0.0.0', function () {
        const str = '10.0.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([10, 0, 0, 0]);
    });

    it('192.168.0.0', function () {
        const str = '192.168.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([192, 168, 0, 0]);
    });

    it('172.16.0.0', function () {
        const str = '172.16.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([172, 16, 0, 0]);
    });

    it('10.10.0.0', function () {
        const str = '10.10.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([10, 10, 0, 0]);
    });

    it('192.0.2.0', function () {
        const str = '192.0.2.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([192, 0, 2, 0]);
    });

    it('203.0.113.0', function () {
        const str = '203.0.113.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([203, 0, 113, 0]);
    });

    it('198.51.100.0', function () {
        const str = '198.51.100.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([198, 51, 100, 0]);
    });

    it('172.18.0.0', function () {
        const str = '172.18.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([172, 18, 0, 0]);
    });

    it('10.20.30.0', function () {
        const str = '10.20.30.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([10, 20, 30, 0]);
    });

    it('192.168.1.0', function () {
        const str = '192.168.1.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([192, 168, 1, 0]);
    });

    it('172.31.0.0', function () {
        const str = '172.31.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([172, 31, 0, 0]);
    });

    it('10.100.0.0', function () {
        const str = '10.100.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([10, 100, 0, 0]);
    });

    it('203.0.123.0', function () {
        const str = '203.0.123.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([203, 0, 123, 0]);
    });

    it('198.18.0.0', function () {
        const str = '198.18.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([198, 18, 0, 0]);
    });

    it('172.19.0.0', function () {
        const str = '172.19.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([172, 19, 0, 0]);
    });

    it('10.50.0.0', function () {
        const str = '10.50.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([10, 50, 0, 0]);
    });

    it('192.168.2.0', function () {
        const str = '192.168.2.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([192, 168, 2, 0]);
    });

    it('172.17.0.0', function () {
        const str = '172.17.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([172, 17, 0, 0]);
    });

    it('10.200.0.0', function () {
        const str = '10.200.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([10, 200, 0, 0]);
    });

    it('192.168.3.0', function () {
        const str = '192.168.3.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([192, 168, 3, 0]);
    });

    it('10.0.0.0/24', function () {
        const str = '10.0.0.0/24';
        const res = validateIPv4(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should not contain mask part`,
        });
    });

    it('10.0.0..0', function () {
        const str = '10.0.0..0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `each section in ${str} should be one valid number`,
        });
    });

    it('10.256.0.0', function () {
        const str = '10.256.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `each part in ${str} should between 0-255`,
        });
    });

    it('10.10.0', function () {
        const str = '10.10.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain 4 parts`,
        });
    });

    it('10.10.0.0.0', function () {
        const str = '10.10.0.0.0';
        const res = validateIPv4(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain 4 parts`,
        });
    });

    it('10.10.0.a', function () {
        const str = '10.10.0.a';
        const res = validateIPv4(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} may contain non-numberic character`,
        });
    });
});

tests/cidr.test.ts
import { validateCIDR } from '../src/utils/common-utils';

describe('cidr validation', function () {
    it('10.0.0.0/24', function () {
        const str = '10.0.0.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[10, 0, 0, 0], 24]);
    });

    it('192.168.0.0/16', function () {
        const str = '192.168.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[192, 168, 0, 0], 16]);
    });

    it('172.16.0.0/12', function () {
        const str = '172.16.0.0/12';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[172, 16, 0, 0], 12]);
    });

    it('10.10.0.0/16', function () {
        const str = '10.10.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[10, 10, 0, 0], 16]);
    });

    it('192.0.2.0/24', function () {
        const str = '192.0.2.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[192, 0, 2, 0], 24]);
    });

    it('203.0.113.0/24', function () {
        const str = '203.0.113.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[203, 0, 113, 0], 24]);
    });

    it('198.51.100.0/24', function () {
        const str = '198.51.100.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[198, 51, 100, 0], 24]);
    });

    it('172.18.0.0/16', function () {
        const str = '172.18.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[172, 18, 0, 0], 16]);
    });

    it('10.20.30.0/24', function () {
        const str = '10.20.30.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[10, 20, 30, 0], 24]);
    });

    it('192.168.1.0/24', function () {
        const str = '192.168.1.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[192, 168, 1, 0], 24]);
    });

    it('172.31.0.0/16', function () {
        const str = '172.31.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[172, 31, 0, 0], 16]);
    });

    it('10.100.0.0/16', function () {
        const str = '10.100.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[10, 100, 0, 0], 16]);
    });

    it('203.0.123.0/24', function () {
        const str = '203.0.123.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[203, 0, 123, 0], 24]);
    });

    it('198.18.0.0/15', function () {
        const str = '198.18.0.0/15';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[198, 18, 0, 0], 15]);
    });

    it('172.19.0.0/16', function () {
        const str = '172.19.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[172, 19, 0, 0], 16]);
    });

    it('10.50.0.0/16', function () {
        const str = '10.50.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[10, 50, 0, 0], 16]);
    });

    it('192.168.2.0/24', function () {
        const str = '192.168.2.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[192, 168, 2, 0], 24]);
    });

    it('172.17.0.0/16', function () {
        const str = '172.17.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[172, 17, 0, 0], 16]);
    });

    it('10.200.0.0/16', function () {
        const str = '10.200.0.0/16';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[10, 200, 0, 0], 16]);
    });

    it('192.168.3.0/24', function () {
        const str = '192.168.3.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(true);
        expect(res.val).toStrictEqual([[192, 168, 3, 0], 24]);
    });

    it('10.0.0.0', function () {
        const str = '10.0.0.0';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain mask part`,
        });
    });

    it('10.0.0..0', function () {
        const str = '10.0.0..0';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `each section in ${str} should be one valid number`,
        });
    });

    it('10.256.0.0/24', function () {
        const str = '10.256.0.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `each part in ${str} should between 0-255`,
        });
    });

    it('10.10.0/24', function () {
        const str = '10.10.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain 4 parts`,
        });
    });

    it('10.10.0.0.0/24', function () {
        const str = '10.10.0.0.0/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain 4 parts`,
        });
    });

    it('10.10.0.0/33', function () {
        const str = '10.10.0.0/33';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `mask in ${str} shuold between 1-32`,
        });
    });

    it('10.10.0.a/24', function () {
        const str = '10.10.0.a/24';
        const res = validateCIDR(str);
        expect(res.ok).toBe(false);
        expect(res.val).toStrictEqual({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} may contain non-numberic character`,
        });
    });
});

src/utils/cidr-utils.ts
export type CIDRDetail = {
    range: string,
    netmask: string,
    wildcardBits: string,
    firstIp: string,
    firstIpInt: number,
    lastIp: string,
    lastIpInt: number,
    totalHost: number,
}

function intToIPv4(ipInt: number): string {
    const arr = [
        ((ipInt >> 24) & 255),
        ((ipInt >> 16) & 255),
        ((ipInt >> 8) & 255),
        ipInt & 255,
    ]
    return arr.join(".")
}


export function splitCIDR(cidr: CIDR): CIDRDetail {
    const [ip, maskInt] = cidr;
    const range = `${ip.join(".")}/${maskInt}`;

    const wildcard = (0x1 << (32 - maskInt)) - 1;
    const mask32 = 0xffffffff;
    const netmask = (~wildcard) & mask32;

    const ipInt = ip.reduce((prev, curr) => (prev << 8) + curr, 0)

    // just in case, coerce to Uint32
    const firstIp = (ipInt & netmask) >>> 0;
    // just in case, coerce to Uint32
    const lastIp = (wildcard | firstIp) >>> 0;

    return {
        range: range,
        netmask: intToIPv4(netmask),
        wildcardBits: intToIPv4(wildcard),
        firstIp: intToIPv4(firstIp),
        firstIpInt: firstIp,
        lastIp: intToIPv4(lastIp),
        lastIpInt: lastIp,
        totalHost: lastIp - firstIp + 1,
    }
}


src/utils/validation-utils.ts
import { Ok, Err, Result } from 'ts-results-es';

export const isEmpty = (string: string | null | undefined) => {
    return !(string != null && String(string).length > 0);
};

type ValidRes = IPV4 | CIDR

function validateNetworkInfo(str: string, is_cidr: boolean): Result<ValidRes, IPValidationError> {
    const ipSections: number[] = [];

    let lastNum = -1;
    let hasMask = false;
    for (let i = 0; i < str.length; i++) {
        const char = str.charCodeAt(i);
        // 46 is '.', 47 is '/'
        if (char == 46 || char == 47) {
            if (lastNum == -1) {
                return Err({
                    kind: 'IP_VALIDATION_ERROR',
                    msg: `each section in ${str} should be one valid number`
                });
            }
            ipSections.push(lastNum);
            // reset memo
            lastNum = -1;
            hasMask = hasMask || (char == 47);
            continue;
        }
        const num = char - 48; // 48 is '0'
        if (num < 0 || num > 9) {
            return Err({
                kind: 'IP_VALIDATION_ERROR',
                msg: `${str} may contain non-numberic character`
            });
        }
        lastNum = lastNum == -1 ? num : lastNum * 10 + num;
    }
    const mask: Mask = lastNum;
    if (!is_cidr) {
        ipSections.push(lastNum);
    }

    if (is_cidr && !hasMask) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain mask part`
        });
    }

    if (!is_cidr && hasMask) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should not contain mask part`
        });
    }

    if (ipSections.length != 4) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain 4 parts`
        });
    }
    if (ipSections.filter(num => num < 0 || num > 255).length != 0) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `each part in ${str} should between 0-255`
        });
    }
    if (is_cidr && (mask <= 0 || mask > 32)) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `mask in ${str} shuold between 1-32`
        });
    }
    const ipv4: [number, number, number, number,] = [
        ipSections[0], ipSections[1], ipSections[2], ipSections[3]
    ]

    return Ok(is_cidr ? [ipv4, mask] : ipv4);
}

export function validateCIDR(str: string): Result<CIDR, IPValidationError> {
    const res = validateNetworkInfo(str, true);
    if (res.ok) {
        return Ok(res.val as CIDR);
    }
    return res
}

export function validateIPv4(str: string): Result<IPV4, IPValidationError> {
    const res = validateNetworkInfo(str, false);
    if (res.ok) {
        return Ok(res.val as IPV4);
    }
    return res
}

src/utils/common-utils.ts
import { Ok, Err, Result } from 'ts-results';

export const isEmpty = (string: string | null | undefined) => {
    return !(string != null && String(string).length > 0);
};

type IPV4 = [number, number, number, number]
type Mask = number
type CIDR = [IPV4, Mask]

type ValidRes = IPV4 | CIDR

function validateNetworkInfo(str: string, is_cidr: boolean): Result<ValidRes, IPValidationError> {
    const ipSections: number[] = [];

    let lastNum = -1;
    let hasMask = false;
    for (let i = 0; i < str.length; i++) {
        const char = str.charCodeAt(i);
        // 46 is '.', 47 is '/'
        if (char == 46 || char == 47) {
            if (lastNum == -1) {
                return Err({
                    kind: 'IP_VALIDATION_ERROR',
                    msg: `each section in ${str} should be one valid number`
                });
            }
            ipSections.push(lastNum);
            // reset memo
            lastNum = -1;
            hasMask = hasMask || (char == 47);
            continue;
        }
        const num = char - 48; // 48 is '0'
        if (num < 0 || num > 9) {
            return Err({
                kind: 'IP_VALIDATION_ERROR',
                msg: `${str} may contain non-numberic character`
            });
        }
        lastNum = lastNum == -1 ? num : lastNum * 10 + num;
    }
    const mask: Mask = lastNum;
    if (!is_cidr) {
        ipSections.push(lastNum);
    }

    if (is_cidr && !hasMask) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain mask part`
        });
    }

    if (!is_cidr && hasMask) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should not contain mask part`
        });
    }

    if (ipSections.length != 4) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `${str} should contain 4 parts`
        });
    }
    if (ipSections.filter(num => num < 0 || num > 255).length != 0) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `each part in ${str} should between 0-255`
        });
    }
    if (is_cidr && (mask <= 0 || mask > 32)) {
        return Err({
            kind: 'IP_VALIDATION_ERROR',
            msg: `mask in ${str} shuold between 1-32`
        });
    }
    const ipv4: [number, number, number, number,] = [
        ipSections[0], ipSections[1], ipSections[2], ipSections[3]
    ]

    return Ok(is_cidr ? [ipv4, mask] : ipv4);
}

export function validateCIDR(str: string): Result<CIDR, IPValidationError> {
    const res = validateNetworkInfo(str, true);
    if (res.ok) {
        return Ok(res.val as CIDR);
    }
    return res
}

export function validateIPv4(str: string): Result<IPV4, IPValidationError> {
    const res = validateNetworkInfo(str, false);
    if (res.ok) {
        return Ok(res.val as IPV4);
    }
    return res
}